File: exo2.3.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/wait.h>
#include <errno.h>


/*Ecrire un programme qui cree 5 fils en utilisant une boucle for.
On remarquera que pour que le fils ne cree pas lui-meme plusieurs fils,
il faut interompre la boucle par
un break dans le fils.*/
int main () {

    int i;
    pid_t pid;

    for (i = 1; i <= 5; i++) {
        pid = fork ();
        if (pid == -1) {
            fprintf (stderr, "Erreur dans fork () !\n");
            exit (errno);
        }
        if (pid == 0) {
            fprintf (stdout, "Je suis le fils %d.\n", i);
            break;
        }
    }

    if (pid == 0) {
        exit (0);
    } else {
        for (int j = 1; j <= 5; j++) {
            wait (NULL);
        }
        fprintf (stdout, "Je suis le pere.\n");
    }

    return 0;
}

File: exo2.3.py
import os
import sys

# Programme qui crée 5 fils en utilisant une boucle for
# Le fils ne doit pas créer lui-même plusieurs fils

def main():
    pid = None
    
    for i in range(1, 6):
        pid = os.fork()
        if pid == -1:
            print("Erreur dans fork() !", file=sys.stderr)
            sys.exit(1)
        if pid == 0:
            print(f"Je suis le fils {i}.")
            break  # Sortie de la boucle pour éviter que le fils ne crée d'autres fils
    
    if pid == 0:
        sys.exit(0)
    else:
        # Le père attend ses 5 fils
        for _ in range(5):
            os.wait()
        print("Je suis le pere.")

if __name__ == "__main__":
    main()
